File: 0827-making-a-large-island/0827-making-a-large-island.py
class Solution:
    def __init__(self):
        self.components = [[]]
        self.size = dict()
        self.n = 0
        self.DIR = [1, 0, -1, 0, 1]
        self.grid = [[]]
        
    def isValid(self, r: int, c: int) -> bool:
        return 0 <= r < self.n and 0 <= c < self.n
        
    def dfs(self, r: int, c: int, flag: int) -> None:
        self.components[r][c] = flag
        self.size[flag] += 1
        for d in range(4):
            nr = r + self.DIR[d]
            nc = c + self.DIR[d + 1]
            if self.isValid(nr, nc) and self.grid[nr][nc] and self.components[nr][nc] == -1:
                self.dfs(nr, nc, flag)
        
    def largestIsland(self, grid: List[List[int]]) -> int:
        self.n = len(grid)
        self.grid = grid
        self.components = [[-1 for _ in range(self.n)] for _ in range(self.n)]
        flag = 0
        
        for r in range(self.n):
            for c in range(self.n):
                if self.grid[r][c] and self.components[r][c] == -1:
                    self.size[flag] = 0
                    self.dfs(r, c, flag)
                    print(self.size[flag])
                    flag += 1
                    
        ans = 0
        
        for r in range(self.n):
            for c in range(self.n):
                if not self.grid[r][c]:
                    save = set()
                    for d in range(4):
                        nr = r + self.DIR[d]
                        nc = c + self.DIR[d + 1]
                        if self.isValid(nr, nc) and self.grid[nr][nc] == 1:
                            save.add(self.components[nr][nc])
                            
                    total = 1
                    for flag in save:
                        total += self.size[flag]
                        
                    ans = max(ans, total)
                    
        return self.n * self.n if not ans else ans
                

File: 0827-making-a-large-island/0827-making-a-large-island.cpp
class Solution {
private:
    vector<vector<int>> components, grid;
    unordered_map<int, int> sizes;
    int n;
    
public:
    const int di[4] = {-1, 1, 0, 0};
    const int dj[4] = {0, 0, -1, 1};
    
    bool isValid(int i, int j) {
        return 0 <= i and i < n and 0 <= j and j < n;
    }
    
    int dfs(int i, int j, int color) {
        if (!isValid(i, j)) {
            return 0;
        }
        if (!(grid[i][j] == 1 and components[i][j] == -1)) {
            return 0;
        }
        components[i][j] = color;
        int cnt = 1;
        for (int p = 0; p < 4; ++p) {
            cnt += dfs(i + di[p], j + dj[p], color);
        }
        return cnt;
    }
        
    int largestIsland(vector<vector<int>>& grid) {
        this->grid = grid;
        n = grid.size();
        components.resize(n, vector<int>(n, -1));
        int color = 0;
        
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] and components[i][j] == -1) {
                    sizes[color] = dfs(i, j, color);
                    color++;
                }
            }
        }
        
        int ans = 0;
        
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (!grid[i][j]) {
                    unordered_set<int> setColor;
                    for (int p = 0; p < 4; ++p) {
                        int ni = i + di[p];
                        int nj = j + dj[p];
                        if (!isValid(ni, nj)) {
                            continue;
                        }
                        if (grid[ni][nj]) {
                            setColor.insert(components[ni][nj]);
                        }
                    }
                    
                    int totalSize = 0;
                    
                    for (int color : setColor) {
                        totalSize += sizes[color];
                    }
                    
                    ans = max(ans, totalSize + 1);
                }
            }
        }
        
        return ans == 0 ? n*n : ans;
    }
};